themes/default/content/blog/five-years-of-infrastructure-as-code-part-one/index.md

```diff
@@ -0,0 +1,208 @@
+
+This newer approach has significant benefits. Fault tolerance is built in. Workloads can scale to zero when not in use, saving money. At the same time, they can  scale to new heights during peak times, to ensure good performance and end user experiences. Managing such an environment is easier because we can rely on declarative goal states and get ourselves out of the messy business of dealing with partial failures and upgrades.
+
+For reasons that are probably obvious (and possibly disturbing to some), old-world infrastructure is sometimes referred to as “pets,” while the new-world infrastructure is referred to as “cattle.” The idea here is that each individual pet is special to an individual, whereas with an entire farm of cattle, we focus more on the health of the herd than any one individual cow.
+
+## Beyond Configuration
+
+When describing infrastructure as code earlier, we skipped over some nuances, like how it all began. The first wave of infrastructure as code tools like <a href="https://cfengine.com/" target="_blank" rel="noopener noreferrer">CFEngine<i class="text-sm ml-2 fas fa-external-link-alt"></i></a>, <a href="https://puppet.com/" target="_blank" rel="noopener noreferrer">Puppet<i class="text-sm ml-2 fas fa-external-link-alt"></i></a>, <a href="https://www.chef.io/" target="_blank" rel="noopener noreferrer">Chef<i class="text-sm ml-2 fas fa-external-link-alt"></i></a>, <a href="https://www.ansible.com/" target="_blank" rel="noopener noreferrer">Ansible<i class="text-sm ml-2 fas fa-external-link-alt"></i></a>, and <a href="https://saltproject.io/" target="_blank" rel="noopener noreferrer">SaltStack<i class="text-sm ml-2 fas fa-external-link-alt"></i></a>, were born in the “two VMs and a database” world. The most common task was to configure a server after it is created – even when that creation itself was done manually by an operator clicking in the AWS EC2 or vSphere console – and then to keep it updated by patching it from then onwards. As such, these tools were very much oriented around configuration and pets, compared to modern tools that are more oriented towards provisioning and cattle.
+
+The modern approach generally provisions the broad array of cloud infrastructure resources, from containers to serverless functions, and yes, even VMs, and everything in between. Some tools like AWS CloudFormation and Azure Resource Manager (ARM) are cloud-specific, while others like Pulumi and Terraform are cloud-agnostic. In this world of many moving pieces, tools are focused on provisioning new things rather than configuring and patching. In “code,” you declare a desired state, and the infrastructure as code engine creates, updates, or deletes the necessary infrastructure resources to bring that desired state into being.
```

Comment:
```suggestion
The modern approach generally provisions the broad array of cloud infrastructure resources, from containers to serverless functions, and yes, even VMs, and everything in between. Some tools like [AWS CloudFormation](https://aws.amazon.com/cloudformation/) and [Azure Resource Manager (ARM)](https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/overview) are cloud-specific, while others like Pulumi and [Terraform](https://terraform.io) are cloud-agnostic. In this world of many moving pieces, tools are focused on provisioning new things rather than configuring and patching. In “code,” you declare a desired state, and the infrastructure as code engine creates, updates, or deletes the necessary infrastructure resources to bring that desired state into being.
```